### DISCUSSION TITLE
PrefixUpTo end of line.

### DISCUSSION THREAD

**Discussion by alexito4** [pointfreeco/swift-parsing#206] [Q&A] [answered]
Maybe a silly question, but when writing parsers I often find myself struggling with this specific change

If I want to take the input until the end of the line:

```swift
PrefixUpTo("\n")
```

that works fine but sometimes I want to support the multiple newlines from other OSs. 

From a little exploration of the API I see `Whitespace` so I thought that would be the solution:

```swift
PrefixUpTo(Whitespace(.vertical))
```

But that provokes a compiler error, a very confusing one by the way. It seems that PrefixUpTo doesn't accept other Parsers so this won't work. But the compiler error put me on the road of trying conversions From substring/utf8 (or maybe it was because i encountered more problems with that in previous versions so I immediately thought it was that ^^')

So is there a way to make that prefix work with another parser? Or is there another obvious way I'm missing to parse until the end of the line?

**Comment by stephencelis** [comment, marked as answer]
Hi @alexito4! The main reason `PrefixThrough` and `PrefixUpTo` take sequences and not parsers is that checking a prefix can be guaranteed to have a particular performance profile, but using any arbitrary parser can introduce something that is quite inefficient to run against each element/character of a collection/string.

Depending on your use case and needs for your parser, you could still implement these parsers to take parsers themselves, and if you can benchmark/share and determine that for common cases such parsers are _not_ slow, we could discuss bringing them into the main library.

As for an alternative spelling of the same problem. I believe you can achieve it with the following:

```swift
Parse {
  Prefix { !$0.isNewline }
  Whitespace(.vertical)
}
```

**Reply by alexito4** [reply]
That makes sense Stephen :) It was just that when building the parser in question I was not in the mindset of performance (I rarely am with small CLIs that don't take a big amount of data), so when I wanted to change from "\n" to Whitespace I was a bit stuck. This, I think, has gotten quite better because before I struggled with small refactors that broke the types because some parser was only available for utf8 when I usually use substring, but I haven't had this issue lately.

For now I'm using the Prefix and `isNewline `, but I will keep in mind to make a custom Parser to try how I could combine the prefix and whitespace ones.
